Adding lesson on R and SQL

emhart opened the pull request:
I've moved this PR from the SQL lesson to the R lesson repo, using the same databases as SQL.  It's conditioned on students being familiar with the data from the SQL lessons. This is a first draft of a lesson and I'd love some feedback.  I tried to build off the existing curriculum using the Portal Mammals database.  My major goals were:

* Show students how to connect to an existing database
* Use `paste()` and R variables to construct queries and have integration with db queries and R variables
* Show students how to build databases using R as a workflow pipeline.

Is this too simple?  Should I have more examples with more complex queries?  Should I construct a more complicated database?  Other thoughts about how I could improve this lesson?

naupaka commented on 06-r-and-sql.Rmd:
```diff
@@ -0,0 +1,144 @@
+
+>  **Challenge**
+>
+> Write a query that gets counts of genus by plot type. If this seems to advanaced, then get counts of genus by plot_id (eliminates of of the joins)
```
typos: advanaced -> advanced and eliminates of -> eliminates one

emhart replied:
fixed typo

naupaka commented on 06-r-and-sql.Rmd:
````diff
@@ -0,0 +1,144 @@
+All we've done so far is execute the same sorts of queries that can easily be made with a GUI.  Now let's try leveraging some of the power scripted queries.  Imagine you want to know how many rodents were found every other year.  To get this we'll get the range of dates from the database, sequence them by two and make new queries.
+
+```{r Get years of data}
+yearRange <- dbGetQuery(conn,"SELECT min(year),max(year) from surveys")
````
from -> FROM?

emhart replied:
fixed typo

naupaka commented:
I don't think it's too simple (certainly not for a two day workshop where some student will not have seen either R or SQL before starting). I also like that you included adding data to a database/creating a database via R, as this has always been one of the top questions I get when I have taught SQL in SWC.

naupaka commented on 06-r-and-sql.Rmd:
```diff
@@ -0,0 +1,144 @@
+
+>  **Challenge**
+>
+> Add a new table to the existing database.  Perhaps it's data about the five different sites, or any other table you'd like. Query your database from SQLite manager just to verify that you're adding data.
```
I might make this last challenge two parts? e.g. 1) add a second table to the database (and maybe verify via SQLite), then 2) query something that requires a join and then plot or calculate some statistic from that joined data. Just a thought--it might help to bring it back full circle to producing analysis products. Or is that too much.

lwasser commented:
hey ted, et al. this is great! i'm teaching SQL tomorrow using python. Maybe we should build a section in the SQL python lesson to parallel your section on creating a DB FROM R. DO you guys agree we should have consistency across languages in the lessons? If anyone wants to have a look, the in progress version is here started by John Gossett.

http://lwasser.github.io/python-ecology/08-working-with-sql

leah

ethanwhite commented:
The lesson looks good to me and I agree with @naupaka's suggestions.

I do think we want coordination between the R and Python material, including for the SQL lesson, so if @emhart and @lwasser can work together to get them coordinated that would be great. That said, I don't see any reason not to bring this in (following the suggested improvements) and then work on synchronizing them.

fmichonneau commented on 06-r-and-sql.Rmd:
```diff
@@ -0,0 +1,144 @@
+Introduction
+---------
+
+Interacting with databases through scripted languages can offer advantages over querying databases via a GUI interface.  A GUI interfaces for your database is easier to use and allows the quick viewing of adhoc queries.  Querying a database with a programatic intreface (in this case R, but it could be any language) are slightly more complicated.  However the trade-off is that data manipulations a ...
```
A GUI interfaces --> A GUI interface

fmichonneau replied:
intreface -> interface

fmichonneau replied:
are not lost --> are not lost and can be reproduced later? or something along these lines.

emhart replied:
Fixed typo and language.

fmichonneau commented on 06-r-and-sql.Rmd:
```diff
@@ -0,0 +1,144 @@
+Connecting R to sqlite databases
+---------
+
+R can connect to databases through a number of libraries.  In our case we we will use [RSQLite](http://cran.r-project.org/web/packages/RSQLite/index.html) to connect to existing SQLite3 databases.  However you should be able to connect on almost any database in R via JDBC[http://cran.r-project.org/web/packages/RJDBC/index.html](http://cran.r-project.org/web/packages/RJDBC/index.html) or [ODBC](ht ...
```
libraries --> packages?

emhart replied:
Changed libraries to packages

fmichonneau commented on 06-r-and-sql.Rmd:
````diff
@@ -0,0 +1,144 @@
+```{r connect}
+library(RSQLite)
+## Set dbname and driver out of convenience
+myDB <- "~/Downloads/portal_mammals.sqlite"
````
maybe just the file name here? as people may not have downloaded it in their download folder.

emhart replied:
Set to just be filenames.

fmichonneau commented on 06-r-and-sql.Rmd:
```diff
@@ -0,0 +1,144 @@
+q <- "SELECT d.plot_type , c.genus, count(*)
+FROM
+(SELECT a.genus, b.plot
+FROM species a
```
we don't cover table name aliasing in the lesson, but I think it's a good idea to demonstrate it here.

emhart replied:
I added some text on line 88, however I'm not sure I give it the treatment it deserves.  I'm open to some thoughts.

fmichonneau commented on 06-r-and-sql.Rmd:
```diff
@@ -0,0 +1,144 @@
+Building your workflow
+------
+
+Up until now we been working with an existing data. However we can use R as way to build up databases from existing flat files.  We'll use data we generate, but you could easily do this with flat files that you read into R.  Here we'll simulate a simple file of 'sites' where you've measured snout vent lengths (svn) of lizards.
```
an existing data -> existing data
as way to build -> as a way to build

emhart replied:
fixed these.

fmichonneau commented:
Thanks Ted! This is really great.

My main comment is that for the part about creating the database from scratch, it might be better to recreate the portal database using the csv files instead of generating random data. Workshop participants don't seem to relate to randomly generated data as much as they do with real data. It would also illustrate what we go through during the SQLite lesson and the GUI where we create the database by importing sequentially the 3 CSV files.

lwasser commented:
Hello again. we are teaching an R workshop next week and i'm in charge of SQL. woo hoo! Where are we with this lesson? Have the suggested changes been merged? The good news is i can teach this lesson next week and then will have taught both python and R sql so i can suggest changes to align both lessons! Thanks @emhart for pulling this together! i really like @fmichonneau 's idea of pulling in the csv's to create the DB rather than random data as well. it's more applied and learners can see how a DB could be created (and built upon) using their data.

emhart commented:
@lwasser I just pushed all these changes.  I fixed some typos and I changed the last part of the lesson to basically recreate the portal mammals DB as suggested by @fmichonneau.  I'd love to get your feedback on how the lesson goes if you teach it.  I also tried to add some text about aliases and sub queries, so I realize I actually may have jumped ahead of what's covered in the basic lesson.  After you teach this and get some feedback we should coordinate as @ethanwhite suggests.

fmichonneau commented:
Thanks @emhart !

lwasser commented:
Thanks @emhart !! i'll run through this today and i'll let you know how it goes Wednesday!
We don't actually build a DB in the python lesson so it will be good to align both lessons after i teach it! :)
